tests/model/document.js

```diff
@@ -321,42 +321,158 @@ describe( 'Document', () => {
 			expect( spy.calledOnce ).to.be.true;
 		} );
 
-		it( 'should be fired if there was a change in a document tree in a change block and have a batch as param', () => {
-			doc.createRoot();
+		it( 'should be fired if there was a selection change in an (enqueue)change block', () => {
+			const root = doc.createRoot();
+			const spy = sinon.spy();
+
+			root._appendChild( new Text( 'foo' ) );
+
+			doc.on( 'change', spy );
+
+			model.change( writer => {
+				writer.setSelection( Range.createFromParentsAndOffsets( root, 2, root, 2 ) );
+			} );
+
+			expect( spy.calledOnce ).to.be.true;
+		} );
+
+		it( 'should not be fired if writer was used on non-document tree', () => {
 			const spy = sinon.spy();
 
 			doc.on( 'change', ( evt, batch ) => {
 				spy();
 				expect( batch ).to.be.instanceof( Batch );
 			} );
 
-			model.enqueueChange( writer => {
+			model.change( writer => {
+				const docFrag = writer.createDocumentFragment();
+				writer.insertText( 'foo', docFrag, 0 );
+			} );
+
+			expect( spy.calledOnce ).to.be.false;
+		} );
+	} );
+
+	describe( 'event change:data', () => {
+		it( 'should be fired if there was a change in a document tree in a change block and have a batch as a param', () => {
+			doc.createRoot();
+			const spy = sinon.spy();
+
+			doc.on( 'change:data', ( evt, batch ) => {
+				spy();
+				expect( batch ).to.be.instanceof( Batch );
+			} );
+
+			model.change( writer => {
 				writer.insertText( 'foo', doc.getRoot(), 0 );
 			} );
 
 			expect( spy.calledOnce ).to.be.true;
 		} );
 
-		it( 'should be fired if there was a selection change in an (enqueue)change block', () => {
+		it( 'should be fired before the change event', () => {
 			doc.createRoot();
+
+			const callOrder = [];
+
+			doc.on( 'change:data', () => {
+				callOrder.push( 1 );
+			} );
+
+			doc.on( 'change', () => {
+				callOrder.push( 2 );
+			} );
+
+			model.change( writer => {
+				writer.insertText( 'foo', doc.getRoot(), 0 );
+			} );
+
+			expect( callOrder ).to.deep.equal( [ 1, 2 ] );
```

Comment:
I do not get this test. There are 2 events fires? I believe there is only `'change:data'` fired and you attached two listeners to it (`change` listener will also listen on `change:data`).

Reply:
However, there should be a test that it should fire change:data when text is inserted.

Reply:
I removed the above test. I didn't know the exact implementation and was wrong there 😕